// @tryghost/ghost/build/nestjs/modules/admin-api.module.js
"use strict";
var __decorate = (this && this.__decorate) || function (decorators, target, key, desc) {
    var c = arguments.length, r = c < 3 ? target : desc === null ? desc = Object.getOwnPropertyDescriptor(target, key) : desc, d;
    if (typeof Reflect === "object" && typeof Reflect.decorate === "function") r = Reflect.decorate(decorators, target, key, desc);
    else for (var i = decorators.length - 1; i >= 0; i--) if (d = decorators[i]) r = (c < 3 ? d(r) : c > 3 ? d(target, key, r) : d(target, key)) || r;
    return c > 3 && r && Object.defineProperty(target, key, r), r;
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.AdminAPIModule = void 0;
const common_1 = require("@nestjs/common");
const example_controller_1 = require("../../http/admin/controllers/example.controller");
const example_service_1 = require("../../core/example/example.service");
const example_repository_in_memory_1 = require("../../db/in-memory/example.repository.in-memory");
let AdminAPIModule = class AdminAPIModule {
};
exports.AdminAPIModule = AdminAPIModule;
exports.AdminAPIModule = AdminAPIModule = __decorate([
    (0, common_1.Module)({
        controllers: [example_controller_1.ExampleController],
        exports: [example_service_1.ExampleService],
        providers: [
            example_service_1.ExampleService,
            {
                provide: 'ExampleRepository',
                useClass: example_repository_in_memory_1.ExampleRepositoryInMemory
            }
        ]
    })
], AdminAPIModule);
