// @tryghost/ghost/build/nestjs/modules/app.module.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.AppModule = void 0;
const core_1 = require("@nestjs/core");
const admin_api_module_1 = require("./admin-api.module");
const not_found_fallthrough_filter_1 = require("../filters/not-found-fallthrough.filter");
const example_listener_1 = require("../../listeners/example.listener");
const admin_api_authentication_guard_1 = require("../guards/admin-api-authentication.guard");
const permissions_guard_1 = require("../guards/permissions.guard");
const location_header_interceptor_1 = require("../interceptors/location-header.interceptor");
const global_exception_filter_1 = require("../filters/global-exception.filter");
class AppModuleClass {
}
exports.AppModule = {
    global: true,
    module: AppModuleClass,
    imports: [admin_api_module_1.AdminAPIModule],
    exports: [],
    controllers: [],
    providers: [
        example_listener_1.ExampleListener,
        {
            provide: core_1.APP_FILTER,
            useClass: global_exception_filter_1.GlobalExceptionFilter
        }, {
            provide: core_1.APP_FILTER,
            useClass: not_found_fallthrough_filter_1.NotFoundFallthroughExceptionFilter
        }, {
            provide: core_1.APP_GUARD,
            useClass: admin_api_authentication_guard_1.AdminAPIAuthentication
        }, {
            provide: core_1.APP_GUARD,
            useClass: permissions_guard_1.PermissionsGuard
        }, {
            provide: core_1.APP_INTERCEPTOR,
            useClass: location_header_interceptor_1.LocationHeaderInterceptor
        }
    ]
};
